### PR TITLE
Abide by minimum buying quantity

### PR DESCRIPTION
**Pull request by thierrymarianne**
| Questions | Answers |
| --- | --- |
| Branch? | develop |
| Description? | Apply minimum product quantity in starter theme and classic  theme |
| Type? | bug fix |
| Category? | FO |
| BC breaks? | no |
| Deprecations? | no |
| Fixed ticket? | http://forge.prestashop.com/browse/BOOM-675 |
| How to test? | Update mininal quantity of product with or without combination and check its addition to a cart considering declare minimal quantities |

### CODE REVIEW

File: controllers/front/ProductController.php [hunk 1]
```diff
@@ -318,6 +323,14 @@ public function displayAjaxQuickview()
     public function displayAjaxRefresh()
     {
         $product_for_template = $this->getTemplateVarProduct();
+        $product_attribute_id = $product_for_template['id_product_attribute'];
+
+        if ($this->combinations) {
+            $mininal_product_combination_quantity = intval($this->combinations[$product_attribute_id]['minimal_quantity']);
```
Comment (maximebiloe) [inline controllers/front/ProductController.php]: A little typo here. `mininal`? ^^
Also, for the new variable name, we should use camelCase

Comment (thierrymarianne) [inline controllers/front/ProductController.php]: Thanks for the typo! I would use camelcase but the whole file would look inconsistent.

Comment (maximebiloe) [inline controllers/front/ProductController.php]: Yeah I know, we need to fix our files

**Comment by vincentbz** [comment]
Tested and ok, thanks !
We saw a problem but it's about the quantity selector on Classic that brings back the quantity to the minimum if we click too fast (more than one click per second)

**Comment by mickaelandrieu** [comment]
@thierrymarianne to be rebased :)

File: controllers/front/ProductController.php [hunk 2]
```diff
@@ -326,30 +333,55 @@ public function displayAjaxRefresh()
...
+     * @param $product
+     * @return int
+     */
+    protected function getMinimalProductOrDeclinationQuantity($product)
+    {
+        $productAttributeId = $product['id_product_attribute'];
+        $minimalProductQuantity = 1;
+
+        if ($this->combinations) {
+            $minimalCombinationProductQuantity = intval($this->combinations[$productAttributeId]['minimal_quantity']);
```
Comment (kpodemski) [inline controllers/front/ProductController.php]: intval? :confused:

File: controllers/front/ProductController.php [hunk 3]
```diff
@@ -326,30 +333,55 @@ public function displayAjaxRefresh()
...
+            if ($minimalCombinationProductQuantity) { // Ensure the minimal product combination quantity is not 0;
+                $minimalProductQuantity = $minimalCombinationProductQuantity;
+            }
+        } elseif (array_key_exists('minimal_quantity', $product)) {
```
Comment (kpodemski) [inline controllers/front/ProductController.php]: i think we had `else if` everywhere in PrestaShop and it was treated as an standard

File: controllers/front/ProductController.php [hunk 4]
```diff
@@ -815,6 +848,59 @@ public function getTemplateVarProduct()
...
+    /**
+     * @param $combinationId
+     * @return null|ProductController
+     */
+    public function findProductCombinationById($combinationId)
+    {
+        $foundCombination = null;
+        $combinations = $this->product->getAttributesGroups($this->context->language->id);
+        foreach ($combinations as $combination) {
+            if (intval($combination['id_product_attribute']) === $combinationId) {
```
Comment (kpodemski) [inline controllers/front/ProductController.php]: intval?

File: controllers/front/ProductController.php [hunk 5]
```diff
@@ -815,6 +848,59 @@ public function getTemplateVarProduct()
...
+        return $foundCombination;
+    }
+
+    /**
+     * @param $product
+     * @return int
+     */
+    protected function getRequiredQuantity($product)
+    {
+        $requiredQuantity = (int)Tools::getValue('quantity_wanted', $this->getProductMinimalQuantity($product));
```
Comment (kpodemski) [inline controllers/front/ProductController.php]: space after `(int)`

File: themes/classic/templates/catalog/_partials/product-add-to-cart.tpl [hunk 6]
```diff
@@ -3,39 +3,47 @@
...
       </div>
       <div class="add">
         <button class="btn btn-primary add-to-cart" data-button-action="add-to-cart" type="submit" {if !$product.add_to_cart_url}disabled{/if}>
           <i class="material-icons shopping-cart">&#xE547;</i>
     {l s='Add to cart' d='Shop.Theme.Actions'}
         </button>
         {block name='product_availability'}
-         {if $product.show_availability}
-            <span id="product-availability">
+          <span id="product-availability">
```
Comment (kpodemski) [inline themes/classic/templates/catalog/_partials/product-add-to-cart.tpl]: so we'll have an empty div if availability is not enabled?

File: themes/classic/templates/catalog/_partials/product-add-to-cart.tpl [hunk 7]
```diff
@@ -3,39 +3,47 @@
...
+          {/if}
+          </span>
         {/block}
       </div>
     </div>
     <div class="clearfix"></div>
   {/block}
 
   {block name='product_minimal_quantity'}
+    <p class="product-minimal-quantity">
```
Comment (kpodemski) [inline themes/classic/templates/catalog/_partials/product-add-to-cart.tpl]: yet another empty element if there's no minimal quantity

File: controllers/front/ProductController.php [hunk 8]
```diff
@@ -815,6 +848,59 @@ public function getTemplateVarProduct()
...
 
+    /**
+     * @param $product
+     * @return int
+     */
+    protected function getProductMinimalQuantity($product)
+    {
+        $minimal_quantity = 1;
+
+        if ($product['id_product_attribute']) {
```
Comment (kpodemski) [inline controllers/front/ProductController.php]: `$this->product->id_product_attribute` would not be good enough?

File: themes/_core/js/cart.js [hunk 9]
```diff
@@ -34,6 +34,34 @@ $(document).ready(() => {
       var query = $form.serialize() + '&add=1&action=update';
       var actionURL = $form.attr('action');
 
+      let isQuantityInputValid = ($input) => {
+        var validInput = true;
+
+        $input.each((index, input) => {
+          let $input = $(input);
+          let mininalValue = parseInt($input.attr('min'), 10);
```
Comment (kpodemski) [inline themes/_core/js/cart.js]: typo

File: controllers/front/ProductController.php (hunk 3 above)
Comment (thierrymarianne) [inline controllers/front/ProductController.php]: http://www.php-fig.org/psr/psr-2/#5-1-if-elseif-else

File: controllers/front/ProductController.php (hunk 2 above)
Comment (thierrymarianne) [inline controllers/front/ProductController.php]: Casting to int.

File: controllers/front/ProductController.php (hunk 4 above)
Comment (thierrymarianne) [inline controllers/front/ProductController.php]: Same as above

File: themes/classic/templates/catalog/_partials/product-add-to-cart.tpl (hunk 6 above)
Comment (thierrymarianne) [inline themes/classic/templates/catalog/_partials/product-add-to-cart.tpl]: Left as placeholder for replacement to work properly at https://github.com/thierrymarianne/PrestaShop/blob/5b9ce520cb4adeaaa7f43a2282ba6e5a4efcb9f8/themes/_core/js/product.js#L20

File: themes/classic/templates/catalog/_partials/product-add-to-cart.tpl (hunk 7 above)
Comment (thierrymarianne) [inline themes/classic/templates/catalog/_partials/product-add-to-cart.tpl]: Please see above comment

File: controllers/front/ProductController.php (hunk 3 above)
Comment (julienbourdeau) [inline controllers/front/ProductController.php]: There were both at some points but they have all been replaced by `elseif`

File: controllers/front/ProductController.php (hunk 8 above)
Comment (thierrymarianne) [inline controllers/front/ProductController.php]: This product attribute id comes from the request, I'd rather keep it as is.

See https://github.com/thierrymarianne/PrestaShop/blob/485ace4daa02edb740e22511727bc1def901aa0a/controllers/front/ProductController.php#L813

**Comment by maximebiloe** [comment]
Thank you!

**Pull request merged**